src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,275 @@
+        JLOG(ctx.j.warn()) << "Batch: Hashes array size does not match txns.";
+        return temINVALID;
+    }
+
+    if (txns.empty())
+    {
+        JLOG(ctx.j.warn()) << "Batch: txns array empty.";
+        return temARRAY_EMPTY;
+    }
+
+    if (txns.size() > 8)
+    {
+        JLOG(ctx.j.warn()) << "Batch: txns array exceeds 12 entries.";
+        return temARRAY_TOO_LARGE;
+    }
+
+    for (int i = 0; i < txns.size(); ++i)
+    {
+        STTx const stx = STTx{STObject(txns[i])};
+        if (stx.getTransactionID() != hashes[i])
+        {
+            JLOG(ctx.j.warn()) << "Batch: Hashes array does not match txns.";
+            return temINVALID;
+        }
+
+        if (!stx.isFieldPresent(sfTransactionType))
+        {
+            JLOG(ctx.j.warn())
+                << "Batch: TransactionType missing in array entry.";
+            return temINVALID_BATCH;
+        }
+        if (stx.getFieldU16(sfTransactionType) == ttBATCH)
+        {
+            JLOG(ctx.j.warn()) << "Batch: batch cannot have inner batch txn.";
+            return temINVALID_BATCH;
+        }
+
+        auto const innerAccount = stx.getAccountID(sfAccount);
+        if (stx.getFieldU16(sfTransactionType) == ttACCOUNT_DELETE &&
+            innerAccount == outerAccount)
+        {
+            JLOG(ctx.j.warn())
+                << "Batch: inner txn cannot be account delete when inner and "
+                   "outer accounts are the same.";
+            return temINVALID_BATCH;
+        }
+
+        if (innerAccount != outerAccount)
+        {
+            if (!tx.isFieldPresent(sfBatchSigners))
+            {
+                JLOG(ctx.j.warn()) << "Batch: missing batch signers.";
+                return temBAD_SIGNER;
+            }
+
+            if (tx.getFieldArray(sfBatchSigners).end() ==
+                std::find_if(
+                    tx.getFieldArray(sfBatchSigners).begin(),
+                    tx.getFieldArray(sfBatchSigners).end(),
+                    [innerAccount](STObject const& signer) {
+                        return signer.getAccountID(sfAccount) == innerAccount;
+                    }))
+            {
+                JLOG(ctx.j.warn())
+                    << "Batch: inner txn not signed by the right user.";
+                return temBAD_SIGNER;
+            }
+        }
+    }
+
+    if (tx.isFieldPresent(sfBatchSigners))
+    {
+        auto const requireCanonicalSig =
+            ctx.rules.enabled(featureRequireFullyCanonicalSig)
+            ? STTx::RequireFullyCanonicalSig::yes
+            : STTx::RequireFullyCanonicalSig::no;
+        auto const sigResult =
+            ctx.tx.checkBatchSign(requireCanonicalSig, ctx.rules);
+        if (!sigResult)
+        {
+            JLOG(ctx.j.warn()) << "Batch: invalid batch txn signature.";
+            return temBAD_SIGNATURE;
+        }
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+Batch::preclaim(PreclaimContext const& ctx)
+{
+    if (!ctx.view.rules().enabled(featureBatch))
```

Comment:
This is already checked in `preflight` and is unneeded here.

Do we want to check the `preclaim`s of the inner transactions here?

Reply:
What were you thinking we check here? We cannot run the `preclaim` function for the inner txn type like we do for doApply because they might rely on an object that doesnt yet exist.